### PR TITLE
Update TIM Plugin to accept standard XML format

### PR DESCRIPTION
**Pull request by dan-du-car**
# PR Details
## Description
Currently, the TIM plugin only encodes Curve Speed Warning information in a non-standard XML into a TIM message. I would like to have all ITIS codes and text inputs accepted in a standard XML format and encoded into a TIM message. The below XML is a sample standard XML accepted by the TIM plugin
```
<TravelerInformation>
      <msgCnt>1</msgCnt>
      <timeStamp>115549</timeStamp>
      <packetID>000000000023667BAC</packetID>
      <dataFrames>
        <TravelerDataFrame>
          <sspTimRights>0</sspTimRights>
          <frameType>
            <advisory/>
          </frameType>
          <msgId>
            <roadSignID>
              <position>
                <lat>389549775</lat>
                <long>-771491835</long>
                <elevation>390</elevation>
              </position>
              <viewAngle>1111111111111111</viewAngle>
              <mutcdCode>
                <warning/>
              </mutcdCode>
            </roadSignID>
          </msgId>
          <startTime>115549</startTime>
          <duratonTime>1</duratonTime>
          <priority>7</priority>
          <sspLocationRights>0</sspLocationRights>
          <regions>
            <GeographicalPath>
              <anchor>
                <lat>389549775</lat>
                <long>-771491835</long>
                <elevation>390</elevation>
              </anchor>
              <directionality>
                <both/>
              </directionality>
              <closedPath>
                <true/>
              </closedPath>
              <description>
                <geometry>
                  <direction>1111111111111111</direction>
                  <circle>
                    <center>
                      <lat>389549775</lat>
                      <long>-771491835</long>
                      <elevation>390</elevation>
                    </center>
                    <radius>74</radius>
                    <units>
                      <meter/>
                    </units>
                  </circle>
                </geometry>
              </description>
            </GeographicalPath>
          </regions>
          <sspMsgRights1>0</sspMsgRights1>
          <sspMsgRights2>0</sspMsgRights2>
          <content>
            <advisory>
              <SEQUENCE>
                <item>
                  <itis>7186</itis>
                </item>
              </SEQUENCE>
              <SEQUENCE>
                <item>
                  <itis>9476</itis>
                </item>
              </SEQUENCE>
              <SEQUENCE>
                <item>
                  <itis>13569</itis>
                </item>
              </SEQUENCE>
            </advisory>
          </content>
          <url>987654321</url>
        </TravelerDataFrame>
      </dataFrames>
    </TravelerInformation>
```

## Related Issue
https://github.com/usdot-fhwa-OPS/V2X-Hub/issues/502

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[V2XHUB Contributing Guide](https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by vibhubithar** [comment]
Hi thanks for looking into this, i am using https://webapp.connectedvcs.com/tim/# to generate the TIM, which generates an ANS.1 encoded message based on this XML

<MessageFrame>
  <messageId>31</messageId>
  <value>
    <TravelerInformation>
      <msgCnt>1</msgCnt>
      <packetID>000000000032320993</packetID>
      <dataFrames>
        <TravelerDataFrame>
          <sspTimRights>0</sspTimRights>
          <frameType>
            <advisory/>
          </frameType>
          <msgId>
            <roadSignID>
              <position>
                <lat>488894612</lat>
                <long>-1002960644</long>
              </position>
              <viewAngle>0000000000000000</viewAngle>
              <mutcdCode>
                <warning/>
              </mutcdCode>
            </roadSignID>
          </msgId>
          <startYear>2023</startYear>
          <startTime>136656</startTime>
          <duratonTime>32000</duratonTime>
          <priority>5</priority>
          <sspLocationRights>0</sspLocationRights>
          <regions>
            <GeographicalPath>
              <anchor>
                <lat>378894612</lat>
                <long>-1222960644</long>
              </anchor>
              <laneWidth>366</laneWidth>
              <closedPath>
                <true/>
              </closedPath>
              <direction>0000000000000000</direction>
              <description>
                <path>
                  <offset>
                    <xy>
                      <nodes>
                        <NodeXY>
                          <delta>
                            <node-XY5>
                              <x>88</x>
                              <y>2314</y>
                            </node-XY5>
                          </delta>
                        </NodeXY>
                        <NodeXY>
                          <delta>
                            <node-XY5>
                              <x>843</x>
                              <y>153</y>
                            </node-XY5>
                          </delta>
                        </NodeXY>
                        <NodeXY>
                          <delta>
                            <node-XY5>
                              <x>1097</x>
                              <y>-4106</y>
                            </node-XY5>
                          </delta>
                        </NodeXY>
                        <NodeXY>
                          <delta>
                            <node-XY5>
                              <x>-790</x>
                              <y>-235</y>
                            </node-XY5>
                          </delta>
                        </NodeXY>
                        <NodeXY>
                          <delta>
                            <node-XY5>
                              <x>-1150</x>
                              <y>4194</y>
                            </node-XY5>
                          </delta>
                        </NodeXY>
                      </nodes>
                    </xy>
                  </offset>
                </path>
              </description>
            </GeographicalPath>
          </regions>
          <sspMsgRights1>0</sspMsgRights1>
          <sspMsgRights2>0</sspMsgRights2>
          <content>
            <advisory>
              <SEQUENCE>
                <item>
                  <itis>7186</itis>
                </item>
              </SEQUENCE>
              <SEQUENCE>
                <item>
                  <itis>9476</itis>
                </item>
              </SEQUENCE>
              <SEQUENCE>
                <item>
                  <itis>13569</itis>
                </item>
              </SEQUENCE>
            </advisory>
          </content>
        </TravelerDataFrame>
      </dataFrames>
    </TravelerInformation>
  </value>
</MessageFrame>

File: src/v2i-hub/TimPlugin/src/TimPlugin.cpp [hunk 1]
```diff
@@ -239,92 +209,155 @@ bool TimPlugin::TimDuration()
...
+		try
+		{
+			std::stringstream ss;
+			ss << in.rdbuf();
+			in.close();
+
+			tmx::message_container_type container;
+			container.load<XML>(ss);
+			TimMsg->set_contents(container.get_storage().get_tree());
+			PLOG(logINFO) << "Loaded MapFile and updated TIM." << std::endl;
```
[the remark this answers is not in the document]
Comment (dan-du-car) [reply src/v2i-hub/TimPlugin/src/TimPlugin.cpp:300]: The same MAP file we upload from the Plugin configure parameter. The XML in the map file is the same as the POST request XML.
```
<TravelerInformation>
      <msgCnt>1</msgCnt>
      <timeStamp>115549</timeStamp>
      <packetID>000000000023667BAC</packetID>
      <dataFrames>
        <TravelerDataFrame>
          <sspTimRights>0</sspTimRights>
          <frameType>
            <advisory/>
          </frameType>
          <msgId>
            <roadSignID>
              <position>
                <lat>389549775</lat>
                <long>-771491835</long>
                <elevation>390</elevation>
              </position>
              <viewAngle>1111111111111111</viewAngle>
              <mutcdCode>
                <warning/>
              </mutcdCode>
            </roadSignID>
          </msgId>
          <startTime>115549</startTime>
          <duratonTime>1</duratonTime>
          <priority>7</priority>
          <sspLocationRights>0</sspLocationRights>
          <regions>
            <GeographicalPath>
              <anchor>
                <lat>389549775</lat>
                <long>-771491835</long>
                <elevation>390</elevation>
              </anchor>
              <directionality>
                <both/>
              </directionality>
              <closedPath>
                <true/>
              </closedPath>
              <description>
                <geometry>
                  <direction>1111111111111111</direction>
                  <circle>
                    <center>
                      <lat>389549775</lat>
                      <long>-771491835</long>
                      <elevation>390</elevation>
                    </center>
                    <radius>74</radius>
                    <units>
                      <meter/>
                    </units>
                  </circle>
                </geometry>
              </description>
            </GeographicalPath>
          </regions>
          <sspMsgRights1>0</sspMsgRights1>
          <sspMsgRights2>0</sspMsgRights2>
          <content>
            <advisory>
              <SEQUENCE>
                <item>
                  <itis>7186</itis>
                </item>
              </SEQUENCE>
              <SEQUENCE>
                <item>
                  <itis>9476</itis>
                </item>
              </SEQUENCE>
              <SEQUENCE>
                <item>
                  <itis>13569</itis>
                </item>
              </SEQUENCE>
            </advisory>
          </content>
          <url>987654321</url>
        </TravelerDataFrame>
      </dataFrames>
    </TravelerInformation>
```
The above sample XML is also in the PR description.

File: src/v2i-hub/TimPlugin/src/TimPlugin.cpp [hunk 2]
```diff
@@ -239,92 +209,155 @@ bool TimPlugin::TimDuration()
...
-				//cout<<"TimPlugin:: isMAPfileNEW  "<<_isMapFileNew<<endl;
 				if (_isMapFileNew)
-				{
+				{					
+					PLOG(logINFO)<<"TimPlugin:: isMAPfileNEW  "<<_isMapFileNew<<endl;
+					//reset map update indicator
+					_isMapFileNew = false;
+					//Update the TIM message with XML from map file
+					_timMsgPtr = std::make_shared<TimMessage>();
+					_isTimLoaded = LoadTim(_timMsgPtr, _mapFile.c_str());
```
[the remark this answers is not in the document]
Comment (dan-du-car) [reply src/v2i-hub/TimPlugin/src/TimPlugin.cpp:332]: The XML is the above. The name convention is a bit odd, but yeah, it the a tmp file that has the TIM data.

File: src/v2i-hub/TimPlugin/src/TimPlugin.h [hunk 3]
```diff
@@ -104,7 +105,17 @@ class TimPlugin: public PluginClient {
 	void OnStateChange(IvpPluginState state);
 
 	bool TimDuration();
```
Comment (paulbourelly999) [inline src/v2i-hub/TimPlugin/src/TimPlugin.h:107]: Is this method used anywhere in the new plugin. If not we should remove it and remove the start and end time configuration parameters

File: src/v2i-hub/TimPlugin/manifest.json [hunk 4]
```diff
@@ -25,12 +25,12 @@
 		},
 		{
 			"key":"Start_Broadcast_Date",
-			"default":"01-01-2019",
```
[the remark this answers is not in the document]
Comment (dan-du-car) [reply src/v2i-hub/TimPlugin/manifest.json:28]: It is not used

File: src/v2i-hub/TimPlugin/src/TimPlugin.h (hunk 3 above)
Comment (dan-du-car) [reply src/v2i-hub/TimPlugin/src/TimPlugin.h]: removed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/508/commits/c781252ca54d7d7f6442ee975c055d79f6bd19dc

File: src/v2i-hub/TimPlugin/src/TimPlugin.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/v2i-hub/TimPlugin/src/TimPlugin.cpp]: Can we move the location of the file to somewhere else, which is permission controlled. For example somewhere in the v2x-hub/TimPlugin directory. See the sonar scanner security hot spot code smell.

[the remark this answers is not in the document]
Comment (paulbourelly999) [reply src/v2i-hub/TimPlugin/src/TimPlugin.cpp]: https://sonarcloud.io/project/security_hotspots?id=usdot-fhwa-ops_V2X-Hub&branch=timplugin_xml_update&resolved=false&types=SECURITY_HOTSPOT

**Comment by paulbourelly999** [comment]
> looking into this, i am using https://webapp.connectedvcs.com/tim/# to generate the TIM, which generates an ANS.1 encoded message based on this XML

I think in the tool you posted, you have a drop down near the bottom that allows you to generate the TIM message in multiple formats. The format you have posted I think includes the message frame. I believe if you select the message type "TIM" it will provide the XML posted in this PR that does not include the message frame. Other than that I believe the payloads are identical. Below I have a screen shot of the **message type** drown down menu which I believe controls this.  @jwilliammartin or @dan-du-car have you used the TIM web tool and can you confirm this behavior?
image

File: src/v2i-hub/TimPlugin/src/TimPlugin.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dan-du-car) [reply src/v2i-hub/TimPlugin/src/TimPlugin.cpp]: I do not think this TMP_FILE_PATH  variable used.

**Review by jwillmartin — APPROVED** [review]

**Comment by jwillmartin** [comment]
@paulbourelly999 I tested a couple different versions of the message that included and excluded different optional fields from the standard. I also used messages created by the the ISD Creator Tool. All worked well.

**Pull request merged**